src/main/scala/chiseltest/internal/BackendInterface.scala

```diff
@@ -2,111 +2,93 @@
 
 package chiseltest.internal
 
-import chiseltest.Region
 import chisel3._
 
-import scala.collection.mutable
+import scala.math.BigInt
 
-trait AbstractTesterThread
+/** SimulatorInterface usages in ThreadedBackend. */
+trait BackendInterface { this: ThreadedBackend =>
 
-class TesterThreadList(protected val elts: Seq[AbstractTesterThread]) {
-  def toSeq(): Seq[AbstractTesterThread] = elts
-
-  def join() {
-    Context().backend.doJoin(elts, None)
-  }
-
-  def joinAndStep(clock: Clock) {
-    Context().backend.doJoin(elts, Some(clock))
-  }
-
-  def ++(others: TesterThreadList): TesterThreadList = {
-    new TesterThreadList(elts ++ others.elts)
-  }
-
-  val fork: ForkBuilder = new ForkBuilder(None, None, elts)
-}
-
-/** Common interface definition for tester backends. Internal API.
-  */
-trait BackendInterface {
+  def getClock(clk: Clock): Boolean =
+    simulatorInterface.peek(dataNames(clk)) match {
+      case Some(x) if x == BigInt(1) => true
+      case _                         => false
+    }
 
   /** Writes a value to a clock.
     */
-  def pokeClock(signal: Clock, value: Boolean): Unit
+  def pokeClock(signal: Clock, value: Boolean): Unit = {
+    val intValue = if (value) 1 else 0
+    simulatorInterface.poke(dataNames(signal), intValue)
+    logger.debug(s"${resolveName(signal)} <- $intValue")
+  }
 
   /** Read the value of a clock.
     */
-  def peekClock(signal: Clock): Boolean
+  def peekClock(signal: Clock): Boolean = {
+    doPeek(signal, new Throwable)
+    val a = simulatorInterface.peek(dataNames(signal)).getOrElse(BigInt(0))
+    logger.debug(s"${resolveName(signal)} -> $a")
+    a > 0
+  }
 
   /** Writes a value to a writable wire.
     * Throws an exception if write is not writable.
     */
-  def pokeBits(signal: Data, value: BigInt): Unit
+  def pokeBits(signal: Data, value: BigInt): Unit = {
+    doPoke(signal, value, new Throwable)
+    val dataName = dataNames(signal)
+    simulatorInterface.peek(dataName) match {
+      case Some(peekValue) =>
+        if (peekValue != value) {
+          idleCycles.clear()
+        }
+        simulatorInterface.poke(dataNames(signal), value)
+        logger.debug(s"${resolveName(signal)} <- $value")
+      case None =>
+        logger.debug(s"${resolveName(signal)} is eliminated by firrtl, ignore it. ")
```

Comment:
Similarly, would prefer this throws an exception. The exception text can provide diagnostic hints (eg, "this signal doesn't exist, was it eliminated by firrtl?"), but it should fail very noisily.